File: powerpro/public/js/erpnext/project_template.js
// Copyright (c) 2025, Yefri Tavarez and contributors
// For license information, please see license.txt
/* eslint-disable */

{
	function refresh(frm) {
		_set_queries(frm);
		_add_custom_buttons(frm);
		_setup_docfields_table_as_readonly(frm);
	}

	function onload(frm) {
		_dirty_form_if_new_docfields(frm);
	}

	function _set_queries(frm) {
		// Set query for the "user" under the tasks table
		frm.set_query("user", "tasks", function(_, doctype, name) {
			const doc = frappe.get_doc(doctype, name);

			const query = "powerpro.controllers.queries.get_active_users_in_department";
			const filters = {
				department: doc.department,
			};

			return { query, filters };
		});
	}

	function _add_custom_buttons(frm) {
		// const actions_parent = __("Actions");
		const docfields_parent = __("DocFields");

		frappe.run_serially([
			() => {
				const label = __("Syncronizar");
				const action = () => {
					const method = "set_project_docfields";
					const args = {
						for_reload: true,
						with_memory: true
					};
					function callback(response) {
						frappe.show_alert({
							message: __("Campos de projecto sincronizados"),
							indicator: "green",
						});
					}

					frm.call(method, args, callback);
				};

				frm.add_custom_button(label, action, docfields_parent);
			},
			() => {
				const label = __("Resetear");
				const action = () => {
					frappe.confirm(
						__("¿Estás seguro de que deseas resetear los campos de proyecto?"),
						() => _reset_project_docfields(frm), () => {
							frappe.show_alert({
								message: __("Operación cancelada"),
								indicator: "red",
							});
						}
					);
				};

				frm.add_custom_button(label, action, docfields_parent);
			},
		]);
	}

	function _reset_project_docfields(frm) {
		const { doc } = frm;
		for (
			const child of doc.project_docfields
		) {
			child.hidden = false;
			child.reqd = false;
			child.read_only = false;
		}

		frappe.run_serially([ frm.dirty.bind(frm), frm.save.bind(frm) ]);
	}

	function _setup_docfields_table_as_readonly(frm) {
		frm.set_df_property("project_docfields", "cannot_add_rows", true);
		frm.set_df_property("project_docfields", "cannot_delete_rows", true);
	}

	function _dirty_form_if_new_docfields(frm) {
		const { doc } = frm;

		const any = doc
			.project_docfields
			.filter(d => d.__islocal)
			.length
		;

		if (any) {
			frm.dirty();
		}
	}

	frappe.ui.form.on("Project Template", {
		refresh,
		onload,
	});
}

File: powerpro/controllers/project_template.py
# Copyright (c) 2025, Yefri Tavarez and Contributors
# For license information, please see license.txt

import frappe
from frappe.model import no_value_fields

from erpnext.projects.doctype.project_template import project_template


class ProjectTemplate(project_template.ProjectTemplate):
	def onload(self):
		self.set_project_docfields()

	def before_insert(self):
		self.set_project_docfields()

	def set_project_docfields(self):
		if not self.project_docfields:
			for df in get_project_fields():
				self.append("project_docfields", {
					"label": frappe._(df.label, "es"),
					"fieldname": df.fieldname,
				})

def get_project_fields():
	meta = frappe.get_meta("Project")

	# only value fields
	value_fields = [
		field
		for field in meta.fields
		if field.fieldtype not in no_value_fields
	]

	# get reqd fields
	reqd_fields = [
		field
		for field in value_fields
		if field.reqd
	]


	# get read_only fields
	read_only_fields = [
		field
		for field in value_fields
		if field.read_only
	]

	# exclude reqd and read_only fields
	exclude_fields = reqd_fields + read_only_fields

	# get project fields
	return [
		field
		for field in value_fields
		if field not in exclude_fields
	]


@frappe.whitelist()
def get_project_docfields(project_template_id=None):
	if project_template_id:
		template = get_project_template(project_template_id)

		fields = template.project_docfields
	else:
		fields = get_project_fields()

	return [
		{
			"label": field.label,
			"fieldname": field.fieldname,
			"reqd": field.reqd,
			"read_only": field.read_only,
			"hidden": field.hidden,
		} for field in fields
	]


def get_project_template(name):
	doctype= "Project Template"
	return frappe.get_doc(doctype, name)
